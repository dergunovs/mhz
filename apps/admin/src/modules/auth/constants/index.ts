export const API_POST_LOGIN = 'auth/login';
export const API_GET_CHECK_AUTH = 'auth/check';

export const URL_LOGIN = '/';
