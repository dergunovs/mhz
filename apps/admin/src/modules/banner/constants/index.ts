export const URL_BANNER = '/banners';
export const URL_BANNER_CREATE = `${URL_BANNER}/create`;
export const URL_BANNER_EDIT = `${URL_BANNER}/edit`;
