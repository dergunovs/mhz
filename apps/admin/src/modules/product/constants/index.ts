export const URL_PRODUCT = '/products';
export const URL_PRODUCT_CREATE = `${URL_PRODUCT}/create`;
export const URL_PRODUCT_EDIT = `${URL_PRODUCT}/edit`;
