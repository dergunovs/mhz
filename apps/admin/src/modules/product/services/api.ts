import { Ref, ComputedRef } from 'vue';

import { api, IPageQuery, convertParams } from 'mhz-helpers';
import { API_PRODUCT, IProduct, IFilterData, IBaseReply } from 'mhz-contracts';

export async function getProductsApi(query: Ref<IPageQuery | number>) {
  const params = convertParams(query);

  const { data } = await api.get<{ data: IProduct[]; total: number; filters: IFilterData }>(API_PRODUCT, { params });

  return data;
}

export async function getProductApi(id?: ComputedRef<string | string[]>) {
  if (!id?.value) return null;

  const { data } = await api.get<{ data: IProduct }>(`${API_PRODUCT}/${id.value}`);

  return data.data;
}

export async function postProductApi(formData: IProduct) {
  const { data } = await api.post<IBaseReply>(API_PRODUCT, formData);

  return data;
}

export async function updateProductApi(formData: IProduct) {
  const { data } = await api.patch<IBaseReply>(`${API_PRODUCT}/${formData._id}`, formData);

  return data;
}

export async function deleteProductApi(id?: string) {
  if (!id) return null;

  const { data } = await api.delete<IBaseReply>(`${API_PRODUCT}/${id}`);

  return data;
}
