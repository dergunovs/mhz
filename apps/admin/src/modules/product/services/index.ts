import { Ref, ComputedRef } from 'vue';

import { useQuery, useMutation, IPageQuery } from 'mhz-helpers';
import { API_PRODUCT } from 'mhz-contracts';

import {
  getProductsApi,
  getProductApi,
  postProductApi,
  updateProductApi,
  deleteProductApi,
} from '@/product/services/api';

export function getProducts(query: Ref<IPageQuery | number>) {
  return useQuery({ queryKey: [API_PRODUCT, query], queryFn: () => getProductsApi(query) });
}

export function getProduct(id?: ComputedRef<string | string[]>) {
  return useQuery({ queryKey: [API_PRODUCT, id], queryFn: () => getProductApi(id) });
}

export function postProduct(options: object) {
  return useMutation({ mutationKey: [API_PRODUCT], mutationFn: postProductApi, ...options });
}

export function updateProduct(options: object) {
  return useMutation({ mutationKey: [API_PRODUCT], mutationFn: updateProductApi, ...options });
}

export function deleteProduct(options: object) {
  return useMutation({ mutationKey: [API_PRODUCT], mutationFn: deleteProductApi, ...options });
}
