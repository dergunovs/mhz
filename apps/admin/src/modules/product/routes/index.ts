import { URL_PRODUCT } from '@/product/constants';

export const productRoutes = [
  {
    path: URL_PRODUCT,
    name: 'ProductList',
    component: () => import('@/product/pages/ProductListPage.vue'),
  },
  {
    path: `${URL_PRODUCT}/:id`,
    name: 'Product',
    component: () => import('@/product/pages/ProductPage.vue'),
  },
];
