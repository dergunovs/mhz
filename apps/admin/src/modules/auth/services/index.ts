import { Ref } from 'vue';
import { useMutation, useQuery } from '@tanstack/vue-query';

import { ILoginFormData } from '@/auth/interface';
import { API_GET_CHECK_AUTH, API_POST_LOGIN } from '@/auth/constants';

import { api } from '@/common/services/api';

export function postLogin(formData: Ref<ILoginFormData>, options: object) {
  async function fn(): Promise<string | undefined> {
    const { data } = await api.post(API_POST_LOGIN, formData.value);

    return data.token;
  }

  return useMutation({
    mutationKey: [API_POST_LOGIN, formData],
    mutationFn: fn,
    ...options,
  });
}

export function getCheckAuth(options: object) {
  async function fn() {
    const { data } = await api.get(API_GET_CHECK_AUTH);

    return data;
  }

  return useQuery({
    queryKey: [API_GET_CHECK_AUTH],
    queryFn: fn,
    ...options,
  });
}
