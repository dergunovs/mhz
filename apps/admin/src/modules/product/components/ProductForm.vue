<template>
  <form @submit.prevent="props.product?._id ? update() : submit()" :class="$style.form">
    <div :class="$style.row">
      <UiField label="Title" isRequired :error="error('title')">
        <UiInput v-model="formData.title" isFocus />
      </UiField>

      <UiField label="Price" isRequired :error="error('price')">
        <UiInput v-model="formData.price" type="number" />
      </UiField>
    </div>

    <div :class="$style.row">
      <UiField label="Manufacturer" isRequired :error="error('manufacturer')">
        <UiSelect
          v-model="formData.manufacturer"
          :options="allManufacturers"
          @reachedBottom="
            scrollManufacturers(isLoadingManufacturers, setManufacturerPage(manufacturersPage + 1, manufacturersPage))
          "
        />
      </UiField>

      <UiField label="Category" isRequired :error="error('category')">
        <UiSelect v-model="formData.category" :options="categories" />
      </UiField>
    </div>

    <UiCheckbox label="In stock" v-model="formData.isInStock" isRequired :error="error('isInStock')" />

    <UiField label="Description" isRequired :error="error('description')">
      <UiEditor v-model="formData.description" />
    </UiField>

    <ProductFieldsForm
      v-if="currentFields?.length"
      :fields="currentFields"
      :updates="categoryUpdates"
      @update="updateFields"
    />

    <UiUpload
      label="Upload images"
      :files="images"
      isRequired
      :error="error('imageUrls')"
      :extensions="['jpg', 'png']"
      @add="addImage"
      @remove="removeImage"
      @upload="mutateUploadFiles(images)"
    />

    <ImagePreview
      v-if="formData.imageUrls.length"
      :urls="formData.imageUrls"
      isThumb
      @update="updateImages"
      @delete="deleteImage"
    />

    <FormButtons :id="props.product?._id" :isLoading="isLoadingPost || isLoadingUpdate" @delete="handleDelete" />
  </form>
</template>

<script setup lang="ts">
import { ref, computed, onMounted, watch, toRaw } from 'vue';
import { useRouter } from 'vue-router';

import { UiField, UiInput, UiCheckbox, toast, UiEditor, UiSelect, UiUpload } from 'mhz-ui';
import { ICategory, IProduct, IManufacturer, ICategoryField } from 'mhz-types';
import { clone, usePagination, useInfiniteScroll, useValidator, required, useQueryClient } from 'mhz-helpers';
import { API_PRODUCT } from 'mhz-contracts';

import ProductFieldsForm from '@/product/components/ProductFieldsForm.vue';
import ImagePreview from '@/common/components/ImagePreview.vue';
import FormButtons from '@/common/components/FormButtons.vue';

import { URL_PRODUCT } from '@/product/constants';
import { postProduct, updateProduct, deleteProduct } from '@/product/services';
import { getCategories } from '@/category/services';
import { getManufacturers } from '@/manufacturer/services';
import { uploadFiles } from '@/common/services';

interface IProps {
  product?: IProduct;
}

const props = defineProps<IProps>();

const queryClient = useQueryClient();

const router = useRouter();

const formData = ref<IProduct>({
  title: '',
  description: '',
  price: 0,
  isInStock: false,
  imageUrls: [],
  thumbUrls: [],
  category: {} as ICategory,
  manufacturer: {} as IManufacturer,
  fields: [],
});

const categoryUpdates = ref(0);

const currentFields = computed(() => {
  return props.product?._id && !categoryUpdates.value
    ? toRaw(formData.value.fields)
    : toRaw(formData.value.category.fields);
});

const images = ref<File[]>([]);

function addImage(file: File) {
  images.value = [...images.value, file];
}

function removeImage(fileToRemove: File) {
  images.value = images.value.filter((file) => file.name !== fileToRemove.name);
}

function deleteImage(imageToDelete: string, isThumb: boolean) {
  formData.value.imageUrls = formData.value.imageUrls.filter((image) => image !== imageToDelete);

  if (isThumb) {
    const thumbToDelete = `thumb-${imageToDelete}.webp`;

    formData.value.thumbUrls = formData.value.thumbUrls.filter((image) => image !== thumbToDelete);
  }
}

function updateImages(urls: string[]) {
  const thumbs = urls.map((url) => `thumb-${url}.webp`);

  formData.value.imageUrls = [...urls];
  formData.value.thumbUrls = [...thumbs];
}

const { mutate: mutateUploadFiles } = uploadFiles(
  {
    onSuccess: (data: string[]) => {
      const thumbs = data.map((url) => `thumb-${url}.webp`);

      formData.value.imageUrls = [...formData.value.imageUrls, ...data];
      formData.value.thumbUrls = [...formData.value.thumbUrls, ...thumbs];
      images.value = [];
      toast.success('Images added');
    },
  },
  '1200',
  true
);

const { data: categories } = getCategories();

const {
  page: manufacturersPage,
  allData: allManufacturers,
  addData: addManufacturers,
  handleScroll: scrollManufacturers,
} = useInfiniteScroll<IManufacturer>();

const { data: manufacturersData, isLoading: isLoadingManufacturers } = getManufacturers(manufacturersPage);
const { data: manufacturers, setPage: setManufacturerPage } = usePagination(manufacturersData);

watch(
  () => manufacturers.value,
  () => {
    if (manufacturers.value && !isLoadingManufacturers.value) addManufacturers(manufacturers.value);
  }
);

const productId = computed(() => props.product?._id);

const { mutate: mutatePost, isLoading: isLoadingPost } = postProduct({
  onSuccess: async () => {
    await queryClient.refetchQueries({ queryKey: [API_PRODUCT] });
    toast.success('Product added');
    router.push(URL_PRODUCT);
  },
});

const { mutate: mutateUpdate, isLoading: isLoadingUpdate } = updateProduct(productId, {
  onSuccess: async () => {
    await queryClient.refetchQueries({ queryKey: [API_PRODUCT] });
    toast.success('Product updated');
  },
});

const { mutate: mutateDelete } = deleteProduct({
  onSuccess: async () => {
    queryClient.removeQueries({ queryKey: [API_PRODUCT] });
    await queryClient.refetchQueries({ queryKey: [API_PRODUCT] });
    toast.success('Product deleted');
    router.push(URL_PRODUCT);
  },
});

function updateFields(fields: ICategoryField[]) {
  formData.value.fields = [...fields];
}

const rules = computed(() => {
  return {
    title: required,
    description: required,
    price: required,
    isInStock: required,
  };
});

const { error, isValid } = useValidator(formData, rules);

function submit() {
  if (isValid()) mutatePost(formData.value);
}

function update() {
  if (isValid()) mutateUpdate(formData.value);
}

function handleDelete(id: string) {
  mutateDelete(id);
}

onMounted(() => {
  if (props.product) formData.value = clone(props.product);
  if (manufacturers.value) addManufacturers(manufacturers.value);

  watch(
    () => formData.value.category,
    () => {
      if (props.product?._id) formData.value.fields = [];
      categoryUpdates.value++;
    }
  );
});
</script>

<style module lang="scss">
.form {
  display: flex;
  flex-direction: column;
  gap: 24px;
  align-items: flex-start;
}

.row {
  display: flex;
  gap: 24px;
  width: 100%;
}
</style>
