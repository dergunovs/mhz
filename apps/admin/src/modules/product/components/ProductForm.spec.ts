import { DefineComponent } from 'vue';
import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
import { VueWrapper, enableAutoUnmount, flushPromises } from '@vue/test-utils';
import { API_PRODUCT, IBaseReply, ICategoryField, IProduct } from 'mhz-contracts';
import * as helpers from 'mhz-helpers';
import { toast } from 'mhz-ui';
import { dataTest } from 'mhz-helpers';

import ProductForm from './ProductForm.vue';

import { wrapperFactory, mockQueryReply, mockMutationReply, router } from '@/common/test';
import { URL_PRODUCT } from '@/product/constants';
import { PRODUCTS } from '@/product/fixtures';
import { CATEGORIES } from '@/category/fixtures';
import * as productServices from '@/product/services';
import * as manufacturerServies from '@/manufacturer/services';
import * as categoryServies from '@/category/services';
import { MANUFACTURERS } from '@/manufacturer/fixtures';

const spyGetManufacturers = vi.spyOn(manufacturerServies, 'getManufacturers').mockReturnValue(mockQueryReply(PRODUCTS));
const spyGetCategories = vi.spyOn(categoryServies, 'getCategories').mockReturnValue(mockQueryReply(CATEGORIES));

let onSuccessPost: () => void;
const spyMutatePost = vi.fn();

vi.spyOn(productServices, 'postProduct').mockImplementation((options: { onSuccess?: () => void }) => {
  if (options.onSuccess) onSuccessPost = options.onSuccess;

  return mockMutationReply<IBaseReply, IProduct>(spyMutatePost);
});

let onSuccessUpdate: () => void;
const spyMutateUpdate = vi.fn();

vi.spyOn(productServices, 'updateProduct').mockImplementation((options: { onSuccess?: () => void }) => {
  if (options.onSuccess) onSuccessUpdate = options.onSuccess;

  return mockMutationReply<IBaseReply, IProduct>(spyMutateUpdate);
});

let onSuccessDelete: () => void;
const spyMutateDelete = vi.fn();

vi.spyOn(productServices, 'deleteProduct').mockImplementation((options: { onSuccess?: () => void }) => {
  if (options.onSuccess) onSuccessDelete = options.onSuccess;

  return mockMutationReply<IBaseReply, string | undefined>(spyMutateDelete);
});

const spyRefetchQueries = vi.fn();
const spyRemoveQueries = vi.fn();

vi.spyOn(helpers, 'useQueryClient').mockReturnValue({
  refetchQueries: spyRefetchQueries,
  removeQueries: spyRemoveQueries,
} as unknown as helpers.QueryClient);

const spyRouterPush = vi.spyOn(router, 'push');
const spyToastSuccess = vi.spyOn(toast, 'success');

const PRODUCT = PRODUCTS.data[0];
const QUERY_KEY = { queryKey: [API_PRODUCT] };

const productForm = dataTest('product-form');
const productFormTitle = dataTest('product-form-title');
const productFormPrice = dataTest('product-form-price');
const productFormManufacturer = dataTest('product-form-manufacturer');
const productFormCategory = dataTest('product-form-category');
const productFormStock = dataTest('product-form-stock');
const productFormDescription = dataTest('product-form-description');
const productFormFields = dataTest('product-form-fields');
const productFormImages = dataTest('product-form-images');
const productFormButtons = dataTest('product-form-buttons');

let wrapper: VueWrapper;

beforeEach(() => {
  wrapper = wrapperFactory(ProductForm, {
    props: {
      product: PRODUCT,
    },
  });
});

enableAutoUnmount(afterEach);

describe('ProductForm', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(ProductForm)).toBeTruthy();
  });

  it('matches snapshot', async () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('shows product content in fields if props is passed', async () => {
    expect(wrapper.findComponent(productFormTitle).attributes('modelvalue')).toEqual(PRODUCT.title);
    expect(wrapper.findComponent(productFormPrice).attributes('modelvalue')).toEqual(PRODUCT.price.toString());

    expect(wrapper.findComponent<DefineComponent>(productFormManufacturer).vm.$attrs.modelValue).toEqual(
      PRODUCT.manufacturer
    );

    expect(wrapper.findComponent<DefineComponent>(productFormCategory).vm.$attrs.modelValue).toEqual(PRODUCT.category);

    expect(wrapper.findComponent(productFormStock).attributes('modelvalue')).toEqual(PRODUCT.isInStock?.toString());
    expect(wrapper.findComponent(productFormDescription).attributes('modelvalue')).toEqual(PRODUCT.description);

    expect(
      wrapper.findComponent<DefineComponent<{ fields: ICategoryField[] }>>(productFormFields).vm.$props.fields
    ).toEqual(PRODUCT.fields);

    expect(wrapper.findComponent(productFormImages).attributes('urls')).toBe(PRODUCT.imageUrls?.join());
  });

  it('shows image preview if product has image', async () => {
    expect(wrapper.findComponent(productFormImages).exists()).toBe(true);

    await wrapper.setProps({ product: undefined });

    expect(wrapper.findComponent(productFormImages).exists()).toBe(true);
  });

  it('gets manufacturers and categories to fill the select options', async () => {
    expect(spyGetManufacturers).toBeCalledTimes(1);
    expect(spyGetCategories).toBeCalledTimes(1);
  });

  it('updates fields by child component emit', async () => {
    const NEW_FIELDS: ICategoryField[] = [];

    expect(
      wrapper.findComponent<DefineComponent<{ fields: ICategoryField[] }>>(productFormFields).vm.$props.fields
    ).toEqual(PRODUCT.fields);

    wrapper.findComponent<DefineComponent>(productFormFields).vm.$emit('update', NEW_FIELDS);

    expect((wrapper.vm as unknown as { formData: { fields: ICategoryField[] } }).formData.fields).toEqual(NEW_FIELDS);
  });

  it('updates child fields component by category updated', async () => {
    expect((wrapper.vm as unknown as { categoryUpdates: number }).categoryUpdates).toEqual(0);

    await wrapper.findComponent(productFormCategory).setValue([]);

    expect((wrapper.vm as unknown as { categoryUpdates: number }).categoryUpdates).toEqual(1);
  });

  it('updates product', async () => {
    expect(spyMutateUpdate).toBeCalledTimes(0);
    expect(spyRefetchQueries).toBeCalledTimes(0);
    expect(spyToastSuccess).toBeCalledTimes(0);

    const NEW_PRODUCT_TITLE = 'yo';

    await wrapper.findComponent(productFormTitle).setValue(NEW_PRODUCT_TITLE);

    await wrapper.find(productForm).trigger('submit');

    expect(spyMutateUpdate).toBeCalledTimes(1);
    expect(spyMutateUpdate).toBeCalledWith({ ...PRODUCT, title: NEW_PRODUCT_TITLE });

    onSuccessUpdate();

    expect(spyRefetchQueries).toBeCalledTimes(1);
    expect(spyRefetchQueries).toBeCalledWith(QUERY_KEY);

    await flushPromises();

    expect(spyToastSuccess).toBeCalledTimes(1);
  });

  it('deletes product', async () => {
    expect(spyMutateDelete).toBeCalledTimes(0);
    expect(spyRemoveQueries).toBeCalledTimes(0);
    expect(spyRefetchQueries).toBeCalledTimes(0);
    expect(spyToastSuccess).toBeCalledTimes(0);
    expect(spyRouterPush).toBeCalledTimes(0);

    wrapper.findComponent<DefineComponent>(productFormButtons).vm.$emit('delete', PRODUCT._id);

    expect(spyMutateDelete).toBeCalledTimes(1);
    expect(spyMutateDelete).toBeCalledWith(PRODUCT._id);

    onSuccessDelete();

    expect(spyRemoveQueries).toBeCalledTimes(1);
    expect(spyRemoveQueries).toBeCalledWith(QUERY_KEY);

    expect(spyRefetchQueries).toBeCalledTimes(1);
    expect(spyRefetchQueries).toBeCalledWith(QUERY_KEY);

    await flushPromises();

    expect(spyToastSuccess).toBeCalledTimes(1);

    expect(spyRouterPush).toBeCalledTimes(1);
    expect(spyRouterPush).toBeCalledWith(URL_PRODUCT);
  });

  it('creates product', async () => {
    wrapper.unmount();

    expect(spyMutatePost).toBeCalledTimes(0);
    expect(spyRefetchQueries).toBeCalledTimes(0);
    expect(spyToastSuccess).toBeCalledTimes(0);
    expect(spyRouterPush).toBeCalledTimes(0);

    const wrapperWithoutProduct = wrapperFactory(ProductForm, {});

    const NEW_PRODUCT: IProduct = {
      title: 'text',
      price: 1000,
      thumbUrls: ['/thumb-1.jpg', '/thumb-2.jpg'],
      imageUrls: ['/1.jpg', '/2.jpg'],
      manufacturer: MANUFACTURERS.data[0],
      category: CATEGORIES.data[0],
      fields: [],
      description: 'text-2',
      isInStock: true,
    };

    await wrapperWithoutProduct.findComponent(productFormTitle).setValue(NEW_PRODUCT.title);
    await wrapperWithoutProduct.findComponent(productFormPrice).setValue(NEW_PRODUCT.price);
    await wrapperWithoutProduct.findComponent(productFormDescription).setValue(NEW_PRODUCT.description);
    await wrapperWithoutProduct.findComponent(productFormStock).setValue(NEW_PRODUCT.isInStock);

    wrapperWithoutProduct.vm.formData.imageUrls = NEW_PRODUCT.imageUrls;
    wrapperWithoutProduct.vm.formData.thumbUrls = NEW_PRODUCT.thumbUrls;
    wrapperWithoutProduct.vm.formData.manufacturer = NEW_PRODUCT.manufacturer;
    wrapperWithoutProduct.vm.formData.category = NEW_PRODUCT.category;

    await wrapperWithoutProduct.find(productForm).trigger('submit');

    expect(spyMutatePost).toBeCalledTimes(1);
    expect(spyMutatePost).toBeCalledWith(NEW_PRODUCT);

    onSuccessPost();

    expect(spyRefetchQueries).toBeCalledTimes(1);
    expect(spyRefetchQueries).toBeCalledWith(QUERY_KEY);

    await flushPromises();

    expect(spyToastSuccess).toBeCalledTimes(1);

    expect(spyRouterPush).toBeCalledTimes(1);
    expect(spyRouterPush).toBeCalledWith(URL_PRODUCT);
  });
});
