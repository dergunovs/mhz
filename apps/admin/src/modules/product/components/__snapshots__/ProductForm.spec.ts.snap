// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductForm > matches snapshot 1`] = `
"<form class="_form_ba9112" data-test="product-form">
  <div class="_row_ba9112">
    <fieldset label="Title" isrequired=""><input type="text" modelvalue="G.Skill Trident Z5 Neo RGB 2x32 GB DDR5 6000" isfocus="" data-test="product-form-title"></fieldset>
    <fieldset label="Price" isrequired=""><input type="number" modelvalue="31200" data-test="product-form-price"></fieldset>
  </div>
  <div class="_row_ba9112">
    <fieldset label="Manufacturer" isrequired=""><input type="text" modelvalue="[object Object]" options="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]" data-test="product-form-manufacturer"></fieldset>
    <fieldset label="Category" isrequired=""><input type="text" modelvalue="[object Object]" options="[object Object]" data-test="product-form-category"></fieldset>
  </div><input type="checkbox" label="In stock" modelvalue="true" isrequired="" data-test="product-form-stock">
  <fieldset label="Description" isrequired=""><input type="text" data-test="product-form-description"></fieldset>
  <product-fields-form-stub fields="[object Object],[object Object],[object Object],[object Object],[object Object]" updates="0" data-test="product-form-fields"></product-fields-form-stub><input type="file" label="Upload images" isrequired="" extensions="jpg,png">
  <image-preview-stub urls="resized-1692371580194-12.jpg" isthumb="true" data-test="product-form-images"></image-preview-stub>
  <form-buttons-stub id="64df8a7d7ac901ef3e273343" data-test="product-form-buttons"></form-buttons-stub>
</form>"
`;
